### PR TITLE
Kv grad allreduce v2

### PR DESCRIPTION
**Pull request by jlamypoirier**
Pushing my changes to #36 as a separate PR since I can't push to @thomasw21 's branch.
Same as #36, but reduced the diff and improved the comments.

```
Method                            Loss        Runtime     Avg
MHA:                              2.604475    2820 s      564 ms
MQA, TP=1:                        2.641582    2303 s      461 ms 
MQA, before fix:                  3.740988    2538 s      508 ms
MQA, PR 36, first version:        2.640432    2602 s      520 ms
MQA, PR 37:                       2.640481    2528 s      506 ms
MQA, PR 36:                       2.640481    2548 s      510 ms
MQA, PR 39:                       2.640481    2538 s      508 ms
MQA, PR 37, sequence parallel:    2.640036    2654 s      531 ms
MQA, PR 36, sequence parallel:    2.638824    2629 s      526 ms
MQA, PR 39, sequence parallel:    2.638824    2622 s      524 ms
```
So this is identical to PR 36 other than statistical variation on the runtime, same as PR 37 with TP and marginally faster than PR 37 with SP (though it's hard to distinguish the speedup from statistical variation)

### CODE REVIEW

File: megatron/optimizer/distrib_optimizer.py [hunk 1]
```diff
@@ -542,7 +542,11 @@ def reduce_model_grads(self, args, timers):
         timers('backward-embedding-all-reduce').stop()
 
         # All-reduce key-value grads if needed.
-        if args.attention_head_type == "multiquery":
+        if (
+            args.attention_head_type == "multiquery"
+            and mpu.get_tensor_model_parallel_world_size() > 1
```
Comment (thomasw21) [inline megatron/optimizer/distrib_optimizer.py:547]: Maybe for consistency add this if condition in the `allreduce` method?

**Review by thomasw21 — APPROVED** [review]
LGTM! Small nit for consistency as most `allreduce` methods test for process group size within them.

File: megatron/optimizer/distrib_optimizer.py (hunk 1 above)
Comment (jlamypoirier) [reply megatron/optimizer/distrib_optimizer.py:547]: I know the others do it but this is better because the timers have a big overhead.

Comment (thomasw21) [reply megatron/optimizer/distrib_optimizer.py:547]: Actually you only need to sync when it's sequence parallel no?

Comment (thomasw21) [reply megatron/optimizer/distrib_optimizer.py:547]: Hum ... "big overhead" is probably an over statement no? I mean you're storing the time ...

Comment (jlamypoirier) [reply megatron/optimizer/distrib_optimizer.py:547]: It also calls `torch.cuda.synchronize`, that's a lot worse

**Pull request merged**